# Problem 1

*If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.*

*Find the sum of all the multiples of 3 or 5 below 1000.*

Hey, this looks like FizzBuzz. Shouldn't be very hard to solve.

In [1]:
def fizzBuzz(n):
    return sum([i for i in range(n) if i % 3 == 0 or i % 5 == 0])

fizzBuzz(1000)

233168

This is good, but there must be a better way to solve this. After all, every multiple of 15 is divisible by both 3 and 5.

Furthermore, it looks like the sum of the nth set of 15 (using 0-indexing) is equal to 105n + 60.

<table>
    <thead>
        <tr>
            <td>Range</td>
            <td>Values Divisible by 3 or 5</td>
            <td>Sum</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>1-15</td>
            <td>3, 5, 6, 9, 10, 12, 15</td>
            <td>60</td>
        </tr>
        <tr>
            <td>16-30</td>
            <td>18, 20, 21, 24, 25, 27, 30</td>
            <td>165</td>
        </tr>
        <tr>
            <td>31-45</td>
            <td>33, 35, 36, 39, 40, 42, 45</td>
            <td>270</td>
        </tr>
        <tr>
            <td>46-60</td>
            <td>48, 50, 51, 54, 55, 57, 60</td>
            <td>375</td>
        </tr>
    </tbody>
</table>

Let's define a new function using this information.

In [2]:
def fifteens(n):
    fifteens = n//15 
    return sum([60 + 105*(n) for n in range(fifteens)] +[n for n in range(fifteens*15 + 1, 1000) if n % 3 == 0 or n % 5 == 0])

fifteens(1000)

233168

Great! But this is still in linear time. Can we make this run in constant time?

Since the sums of the sets of 15 are an arithmetic sequence, we can use this formula: $$S_{n} = n(\frac{a_{1} + a_{n}}{2})$$

In [3]:
def arithmeticSum(n):
    fifteens = n//15
    return  fifteens * (60 + 105*(fifteens - 1) + 60) // 2 + sum([n for n in range(fifteens*15 + 1, 1000) if n % 3 == 0 or n % 5 == 0])

arithmeticSum(1000)

233168

In [4]:
import time
import numpy as np

def timeDifference(n):
    for fn in (fizzBuzz, fifteens, arithmeticSum):
        times = []
        for i in range(10000):
            start = time.time()
            fn(n)
            times.append(time.time() - start)
        print(fn.__name__ + "({}):".format(n), np.mean(times))

timeDifference(1000)
print()
timeDifference(100000)

fizzBuzz(1000): 7.877552509307861e-05
fifteens(1000): 7.02662467956543e-06
arithmeticSum(1000): 1.4095544815063476e-06

fizzBuzz(100000): 0.007989384388923645
fifteens(100000): 0.000592039179801941
arithmeticSum(100000): 7.001638412475586e-07


Wow! The method using the formula for an arithmetic sum is substantially faster, and the difference in runtime is even more significant for larger ranges (e.g. finding the sum of the numbers divisible by 3 or 5 less than 100,000 instead of 1,000.

Final answer: 233,168